client/src/main/java/io/pravega/client/stream/impl/StreamSegments.java

```diff
@@ -63,66 +74,110 @@ public Segment getSegmentForKey(String key) {
-        NavigableMap<Double, Segment> result = new TreeMap<>();
+    public StreamSegments withReplacementRange(Segment segment, StreamSegmentsWithPredecessors replacementRanges) {
+        SegmentWithRange replacedSegment = findReplacedSegment(segment);
+        verifyReplacementRange(replacedSegment, replacementRanges);
+        NavigableMap<Double, SegmentWithRange> result = new TreeMap<>();
         Map<Long, List<SegmentWithRange>> replacedRanges = replacementRanges.getReplacementRanges();
-        List<SegmentWithRange> replacements = replacedRanges.get(replacedSegment.getSegmentId());
-        Preconditions.checkNotNull(replacements, "Empty set of replacements for: {}", replacedSegment.getSegmentId());
+        List<SegmentWithRange> replacements = replacedRanges.get(segment.getSegmentId());
+        Preconditions.checkNotNull(replacements, "Empty set of replacements for: {}", segment.getSegmentId());
         replacements.sort(Comparator.comparingDouble(SegmentWithRange::getHigh).reversed());
-        Segment lastSegmentValue = null;
-        for (Entry<Double, Segment> existingEntry : segments.descendingMap().entrySet()) { // iterate from the highest key.
-            final Segment existingSegment = existingEntry.getValue();
-            if (existingSegment.equals(lastSegmentValue)) {
-                // last value was the same segment, it can be consolidated.
-                continue;
-            }
+        for (Entry<Double, SegmentWithRange> existingEntry : segments.descendingMap().entrySet()) { // iterate from the highest key.
+            final SegmentWithRange existingSegment = existingEntry.getValue();
             if (existingSegment.equals(replacedSegment)) { // Segment needs to be replaced.
                 // Invariant: The successor segment(s)'s range should be limited to the replaced segment's range, thereby
                 // ensuring that newer writes to the successor(s) happen only for the replaced segment's range.
-                for (SegmentWithRange segmentWithRange : replacements) {                        
-                    result.put(Math.min(segmentWithRange.getHigh(), existingEntry.getKey()), segmentWithRange.getSegment());
+                for (SegmentWithRange segmentWithRange : replacements) {
+                    Double lowerBound = segments.lowerKey(existingEntry.getKey()); // Used to skip over items not in the clients view yet.
+                    if (lowerBound == null || segmentWithRange.getHigh() >= lowerBound) { 
+                        result.put(Math.min(segmentWithRange.getHigh(), existingEntry.getKey()), segmentWithRange);
+                    }
                 }
             } else {
                 // update remaining values.
                 result.put(existingEntry.getKey(), existingEntry.getValue());
             }
-            lastSegmentValue = existingSegment; // update lastSegmentValue to reduce number of entries in the map.
         }
+        removeDuplicates(result);
```

Comment:
Could you give an example of a scenario in which we have duplicates?

Reply:
Documented